### ISSUE TITLE
StructRef initialization stuck while using types.deferred_type

### ISSUE THREAD

**Issue by rpopovici** [numba/numba#6077]
While I was trying to use the deferred_type with StructRefs the execution gets stuck, infinite loop or something

```
import numba as nb
from numba import njit, types, typed, float64, int64
from numba.core import types
from numba.experimental import structref
from numba.core.extending import overload_method

@structref.register
class VectorStructType(types.StructRef):
    def preprocess_fields(self, fields):
        return tuple((name, types.unliteral(typ)) for name, typ in fields)

class VectorStruct(structref.StructRefProxy):
    def __new__(cls, x, y, z):
        return structref.StructRefProxy.__new__(cls, x, y, z)

    @property
    def x(self):
        return VectorStruct_get_x(self)

    @property
    def y(self):
        return VectorStruct_get_y(self)

    @property
    def z(self):
        return VectorStruct_get_z(self)

@njit
def VectorStruct_get_x(self):
    return self.x

@njit
def VectorStruct_get_y(self):
    return self.y

@njit
def VectorStruct_get_z(self):
    return self.z

structref.define_proxy(VectorStruct, VectorStructType, ["x", "y", "z"])

vector_struct_type = VectorStructType(fields=(('x', float64), ('y', float64),
                                              ('z', float64)))

PolygonTypeDeferred = types.deferred_type()
@structref.register
class PolygonStructType(types.StructRef):
    def preprocess_fields(self, fields):
        return tuple([
            ('vertices', types.Optional(types.ListType(vector_struct_type))),
            ('parent', types.Optional(PolygonTypeDeferred)),
            ])

class PolygonStruct(structref.StructRefProxy):
    def __new__(cls, vertices, parent):
        return structref.StructRefProxy.__new__(cls, vertices, parent)

    @property
    def vertices(self):
        return PolygonStruct_get_vertices(self)

    @property
    def parent(self):
        return PolygonStruct_get_parent(self)

@njit
def PolygonStruct_get_vertices(self):
    return self.vertices

@njit
def PolygonStruct_get_parent(self):
    return self.parent

structref.define_proxy(PolygonStruct, PolygonStructType, ["vertices", "parent"])

PolygonType_inst = PolygonStructType(fields=(('vertices', types.Optional(types.ListType(vector_struct_type))), ('parent', types.Optional(PolygonTypeDeferred))))
PolygonTypeDeferred.define(PolygonType_inst)

@overload_method(PolygonStructType, "flip")
def PolygonStruct_flip(self):
    def impl(self):
        if self.vertices is not None:
            ref = self.vertices[0]
            ref.x = -ref.x
    return impl

@njit
def test():
    lst = typed.List.empty_list(vector_struct_type)
    lst.append(VectorStruct(1.0, 2.0, 3.0))
    lst[0].x = 5.0

    print('before')
    p = PolygonStruct(None, None)
    p.flip()
    p.vertices = lst
    p.flip()
    print('after')
    return p

print('start')
y = test()
print(y.vertices[0].x)
print('end')
```

**Comment by stuartarchibald** [comment]
Thanks for the report. This is a `sigabrt` for me:
```
$ python issue6077.py 
start
python: /opt/conda/conda-bld/llvmdev_1591124166996/work/lib/IR/DataLayout.cpp:696: unsigned int llvm::DataLayout::getAlignment(llvm::Type*, bool) const: Assertion `Ty->isSized() && "Cannot getTypeInfo() on a type that is unsized!"' failed.
Aborted
```

**Comment by rpopovici** [comment]
@stuartarchibald I was running it from a jupyter notebook and I got no feedback whatsoever. I thought is was stuck

**Comment by stuartarchibald** [comment]
In jupyter-notebooks everything executes in a separate kernel, what likely happened was that this code caused a SIGABRT and this killed the kernel, but that would have been silent/not obvious unless looked for, so it would have appeared stuck.

**Comment by sklam** [comment]
i think LLVM aborted at creating a LLVM struct containing an undefined opaque struct.

**Comment by rpopovici** [comment]
Can anyone suggest an workaround here until it gets fixed? I tried myself to mess with DeferredType's impmentation without luck so far..

**Comment by rpopovici** [comment]
I found an workaround but now I get a boxing error when I try to instantiate from python mode:
```
TypeError: _numba_box_() missing 2 required positional arguments: 'ty' and 'mi'
```

The workaround
```
PolygonTypeDeferred = types.deferred_type()
@structref.register
class PolygonStructType(types.StructRef):
    pass

PolygonType_inst = PolygonStructType(fields=(('vertices', types.Optional(types.ListType(vector_struct_type))), ('parent', types.Optional(PolygonTypeDeferred))))
PolygonTypeDeferred.define(PolygonType_inst)


def preprocess_fields(self, fields):
    return tuple([
        ('vertices', types.Optional(types.ListType(vector_struct_type))),
        ('parent', types.Optional(PolygonType_inst)),
        ])

# replace preprocess_fields after deferred define
PolygonStructType.preprocess_fields = preprocess_fields
```

**Comment by rpopovici** [comment]
@sklam is there any chance for issue to be fixed in the near future? I am kind of blocked by this and I need to know if I have to rethink my strategy

**Comment by DannyWeitekamp** [comment]
@rpopovici perhaps by now you have moved on but [this](https://numba.discourse.group/t/any-numba-equivalent-for-casting-a-raw-pointer-to-a-structref-dict-list-etc/351/6) could work as a workaround. Admittedly it is a bit of a hack, I've been using this to kind of working around this issue by just passing around meminfos and using types.MemInfoPointer(types.voidptr) as a sort of untyped refcounted pointer placeholder type.

**Comment by rpopovici** [comment]
@DannyWeitekamp thanks for this solution. Type casting is an important feature, not very well supported in numba.. Anyway, for my use case, self-referencing struct types are supported now by this PR #6194 There are examples in the test file about how to use it

**Comment by rpopovici** [comment]
@DannyWeitekamp there is also this draft PR #6209 from @sklam meant to add DeferredType support to structrefs
